## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
rand_arr = np.random.random((20)) * 10
print(rand_arr[(rand_arr > 3) & (rand_arr<8)])

[7.46560826 5.38100956 4.67893492 7.99093252 7.55540033 5.87236373
 5.12012158 7.57248539 3.03847689 3.83751441 3.78995066]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
rand_arr = np.random.random((2,2)) * 10
print(rand_arr)
rand_arr.flat[rand_arr.argmax()] = 0
print(rand_arr)

[[6.06132989 2.55756474]
 [7.67630617 5.09894659]]
[[6.06132989 2.55756474]
 [0.         5.09894659]]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
#Без понятия как тут используется numpy, нашел такое решение с itertools
from itertools import product

arr = np.array([
    [1, 2, 3],
    [3,4,5],
    [6,7,8]
])

b = np.array([
    [1, 2],
    [3,4]
])
#print(np.array(list(product(*arr))))

def cartestian_product(arr):
    #Create cartesian product array
    x = np.array(np.meshgrid(*arr)).T
    #Fix the dimensions by saying to numpy to reshape to (?, len(arr)) and numpy will automatically fit based on the len(arr) 
    return x.reshape(-1, len(arr))

print(cartestian_product(arr))

[[1 3 6]
 [1 4 6]
 [1 5 6]
 [2 3 6]
 [2 4 6]
 [2 5 6]
 [3 3 6]
 [3 4 6]
 [3 5 6]
 [1 3 7]
 [1 4 7]
 [1 5 7]
 [2 3 7]
 [2 4 7]
 [2 5 7]
 [3 3 7]
 [3 4 7]
 [3 5 7]
 [1 3 8]
 [1 4 8]
 [1 5 8]
 [2 3 8]
 [2 4 8]
 [2 5 8]
 [3 3 8]
 [3 4 8]
 [3 5 8]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
A = (np.random.random((8,3)) * 10).astype(int)
B = (np.random.random((2,2)) * 10).astype(int)
print(A)
print(B)

# Using pure python
result = np.zeros(len(A)).astype(bool)
for i in range(len(A)):
    cnt = 0
    for j in range(len(B)):
        for k in range(len(B[0])):
            if B[j][k] in A[i]:
                cnt+=1
                break
    if (cnt == len(B)):
        result[i] = True
print("First result: ")
print(A[result])

#Usin numpy
masks = [np.isin(A, b_row).any(axis = 1) for b_row in B]
masks_all = np.array(masks).all(axis = 0)

print("Second result: ")
print(A[masks_all])
        

[[1 3 7]
 [5 3 6]
 [9 2 9]
 [8 7 1]
 [2 4 4]
 [2 0 1]
 [9 7 5]
 [4 4 7]]
[[4 0]
 [4 2]]
First result: 
[[2 4 4]
 [2 0 1]
 [4 4 7]]
Second result: 
[[2 4 4]
 [2 0 1]
 [4 4 7]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
arr = np.array([
    [1, 1, 1],
    [2, 3, 2],
    [4, 4, 4],
    [5, 6, 7]
])

mask = [row.max() != row.min() for row in arr]
print("Result:\n", arr[mask])

#vectorized
mask = np.ptp(arr, axis = 1) > 0
print("Result:\n", arr[mask])

Result:
 [[2 3 2]
 [5 6 7]]
Result:
 [[2 3 2]
 [5 6 7]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [7]:
arr = np.array([
    [1, 1, 1],
    [1, 3, 2],
    [1, 1, 1],
    [5, 6, 7]
])

print(np.unique(arr, axis=0))

[[1 1 1]
 [1 3 2]
 [5 6 7]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [8]:
X2 = np.array([
    [0, 1],
    [2, 0],
    [3, 4]
])

X3 = np.array([
    [2, 1, 3],
    [4, 5, 6],
    [7, 8, 9]
])


arr = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

#Pure pythoon
n = min(len(arr), len(arr[0]))
res = 1
for i in range(n):
    res *= arr[i][i] if arr[i][i] != 0 else 1
print(res)


# Numpy
d = np.diag(arr)
res = d[d.nonzero()].prod()
print(res)


3
3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [9]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

#Python
from collections import Counter
res = Counter(x) == Counter(y)
print(res)

#Numpy 1
x1, x2 = np.unique(x, return_counts=True)  
y1, y2 = np.unique(y, return_counts=True)

res = np.array_equal(x1, y1) and np.array_equal(x2, y2)

print(res)

#Numpy 2
res = np.array_equal(np.sort(x), np.sort(y))

print(res)

True
True
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [10]:
# your code
arr = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

#Python
maxx = 0
for i in range(1, len(arr)):
    if arr[i-1] == 0 and arr[i] > maxx:
        maxx = arr[i]
        
print(maxx)

#Numpy
zero_indices = np.where(arr[:-1] == 0)[0]
res = arr[zero_indices + 1].max()
print(res)


5
5


Задача 4: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.
Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

<!-- * __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])). -->

In [38]:
# your code
# x = np.array([2, 2, 3, 3, 5, 4,4, 2, 0])
x= (np.random.rand(10) * 3).astype(int)

print(x)
#Python

indices = []
for i in range(len(x)-1):
    if x[i+1] != x[i]:
        indices.append(i)
indices.append(len(x)-1)

ans = []
for i in range(len(indices)):
    start = indices[i-1] if i > 0 else -1
    ans.append((x[indices[i]], indices[i] - start))
print("Python solution:")
for el in ans:
    print(el)


#Numpy
def np_rle(x):
    changes = np.nonzero(np.diff(x) != 0)[0]  + 1 
    starts = np.concatenate(([0], changes))
    ends = np.concatenate((changes, [len(x)]))
    
    lenghts = ends - starts
    values = x[starts]
    
    return list(zip(values, lenghts))


print("Numpy solution")
for el in np_rle(x):
    print(el)


def rle(x):
    values = []
    counts = []
    prev = x[0]
    cnt = 1
    for a in x[1:]:
        if a == prev:
            cnt += 1
        else:
            values.append(prev)
            counts.append(cnt)
            prev = a
            cnt = 1
    # не забываем «добить» последнюю серию
    values.append(prev)
    counts.append(cnt)
    return list(zip(np.array(values), np.array(counts)))

print("chatGPT solution")
for el in rle(x):
    print(el)

print("All 3 answers are the same:")
print(rle(x) == ans and rle(x) == np_rle(x))


[1 2 2 1 2 2 1 1 1 2]
Python solution:
(np.int64(1), 1)
(np.int64(2), 2)
(np.int64(1), 1)
(np.int64(2), 2)
(np.int64(1), 3)
(np.int64(2), 1)
Numpy solution
(np.int64(1), np.int64(1))
(np.int64(2), np.int64(2))
(np.int64(1), np.int64(1))
(np.int64(2), np.int64(2))
(np.int64(1), np.int64(3))
(np.int64(2), np.int64(1))
chatGPT solution
(np.int64(1), np.int64(1))
(np.int64(2), np.int64(2))
(np.int64(1), np.int64(1))
(np.int64(2), np.int64(2))
(np.int64(1), np.int64(3))
(np.int64(2), np.int64(1))
All 3 answers are the same:
True


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [12]:
import time

X = np.random.rand(1000, 3)
Y = np.random.rand(1000, 3)

st = time.time()
res = []
for el1 in X:
    temp = []
    for el2 in Y:
        s = 0
        for i in range(len(el1)):
            s += (el1[i] - el2[i]) ** 2
        temp.append(s ** 0.5)
    res.append(temp)
        
print(time.time() - st)

st = time.time()
diff = X[:, np.newaxis, :] - Y[np.newaxis, :, :]  # shape (m, n, d)
res = dist_matrix = np.linalg.norm(diff, axis=2)
print(time.time() - st)

import scipy
st = time.time()
# print(scipy.spatial.distance.cdist(X, Y))
print(time.time() - st)


0.9786326885223389
0.041169166564941406
2.384185791015625e-05


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [13]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [14]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [15]:
average_calories = calorie_stats.mean()
average_calories

np.float64(106.88311688311688)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [16]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [17]:
median_calories = np.median(calorie_stats)
median_calories

np.float64(110.0)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [18]:
for p in range(101):
    if np.percentile(calorie_stats, p) > 60:
        nth_percentile = p
        break
nth_percentile

4

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [19]:
more_calories = float((calorie_stats > 60).mean().round(3) * 100)
more_calories

96.1

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [20]:
calorie_std = calorie_stats.std()
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

In [21]:
less_then_average = int(60 / average_calories * 100)
less_then_average

56

1. В одной порции CrunchieMunchies всего лишь 60 ккал 
2. В CrunchieMunchies меньше каллорий чем у 96% конкурентов
3. В CrunchieMunchies на 56% меньше каллорий чем в среднем по рынку
